include/libcyphal/platform/single_threaded_executor.hpp

```diff
@@ -113,114 +99,175 @@ class SingleThreadedExecutor : public IExecutor  // NOSONAR cpp:S4963
         return TimePoint{} + std::chrono::duration_cast<Duration>(duration);
     }
 
-    using IExecutor::registerCallback;
-
-protected:
-    CETL_NODISCARD cetl::optional<Callback::Id> appendCallback(Callback::Function&& function) override
+    CETL_NODISCARD Callback::Any registerCallback(Callback::Function&& function) override
     {
         CETL_DEBUG_ASSERT(function, "");
 
-        auto* const new_callback_node = makeCallbackNode(std::move(function));
-        if (nullptr == new_callback_node)
-        {
-            return cetl::nullopt;
-        }
-
-        ++last_callback_id_;
-        const auto new_callback_id = last_callback_id_;
-        new_callback_node->id()    = new_callback_id;
-
-        const std::tuple<CallbackNode*, bool> reg_node_existing = registered_nodes_.search(  //
-            [new_callback_id](const CallbackNode& node) {                                    // predicate
-                return node.compareById(new_callback_id);
-            },
-            [new_callback_node]() { return new_callback_node; });  // factory
+        CallbackNode new_callback_node{*this, std::move(function)};
 
-        (void) reg_node_existing;
-        CETL_DEBUG_ASSERT(!std::get<1>(reg_node_existing), "Callback id collision detected.");
-        CETL_DEBUG_ASSERT(new_callback_node == std::get<0>(reg_node_existing), "Unexpected not the new node.");
+        insertCallbackNode(new_callback_node);
 
-        return cetl::optional<Callback::Id>{new_callback_id};
+        return {std::move(new_callback_node)};
     }
 
-    bool scheduleCallbackById(const Callback::Id                 callback_id,
-                              const TimePoint                    exec_time,
-                              const Callback::Schedule::Variant& schedule) override
+    void scheduleCallback(Callback::Any& callback, const Callback::Schedule::Variant& schedule) override
     {
-        auto* const callback_node_ptr = registered_nodes_.search(  //
-            [callback_id](const CallbackNode& node) {              // predicate
-                return node.compareById(callback_id);
-            });
-        if (nullptr == callback_node_ptr)
+        if (auto* const cb_node = callbackToNode(callback))
         {
-            return false;
+            CETL_DEBUG_ASSERT(this == &cb_node->executor(), "");
+            scheduleCallbackNode(*cb_node, schedule);
         }
-        CallbackNode& callback_node = *callback_node_ptr;
-
-        // Remove previously scheduled node (if any),
-        // and then re/insert the node with updated/given execution time and schedule.
-        //
-        removeIfScheduledNode(callback_node);
-        callback_node.schedule() = schedule;
-        insertScheduledNode(callback_node, exec_time);
-
-        return true;
     }
 
-    void removeCallbackById(const Callback::Id callback_id) override
+protected:
+    static Callback::Handle callbackToHandle(Callback::Any& callback) noexcept
```

Comment:
Maybe I'm missing something obvious but why can't we just use `CallbackNode*` instead? This pointer casting is unsafe.

Reply:
Moved into https://github.com/OpenCyphal-Garage/libcyphal/issues/373